File: planets/bench_pure.py
# Adapted from talk material of:
# N-Body Simulation with Python & Numba
# Dr. Yves J. Hilpisch
# The Python Quants GmbH

import random

import numpy as np
from math import sqrt

nplanets = 1000

planet = []
for i in range(nplanets):
    planet.append([random.gauss(0.0, 1.0) for j in range(3)])
planetv = [[0, 0, 0] for x in planet]


def nbody(planet, planetv):
    nSteps = 15 # In the other code these are 200...
    dt = 0.01
    for _ in range(nSteps):
        for i in range(nplanets):
            Fx = 0.0
            Fy = 0.0
            Fz = 0.0
            for j in range(nplanets):
                if j != i:
                    dx = planet[j][0] - planet[i][0]
                    dy = planet[j][1] - planet[i][1]
                    dz = planet[j][2] - planet[i][2]
                    drSquared = dx * dx + dy * dy + dz * dz
                    drPowerN32 = 1.0 / (drSquared + sqrt(drSquared))
                    Fx += dx * drPowerN32
                    Fy += dy * drPowerN32
                    Fz += dz * drPowerN32
                planetv[i][0] += dt * Fx
                planetv[i][1] += dt * Fy
                planetv[i][2] += dt * Fz
        for i in range(nplanets):
            planet[i][0] += planetv[i][0] * dt
            planet[i][1] += planetv[i][1] * dt
            planet[i][2] += planetv[i][2] * dt
    return planet, planetv


def energy(planets, planetvs):
    e = 0.5 * sum(planetv[0] ** 2 + planetv[1] ** 2 + planetv[2] ** 2 for planetv in planetvs)
    for i in range(nplanets):
        for j in range(nplanets):
            if j != i:
                dx = planet[j][0] - planet[i][0]
                dy = planet[j][1] - planet[i][1]
                dz = planet[j][2] - planet[i][2]
                drSquared = dx * dx + dy * dy + dz * dz
                drPowerN32 = 1.0 / (drSquared + sqrt(drSquared))
                e -= drPowerN32
    return e

e0 = energy(planet, planetv)
planet, planetv = nbody(planet, planetv)
print(f"Error on reference energy: {abs(energy(planet, planetv)/e0)}")

File: planets/bench_pure.ml
let n_planets = 1000

(* This is not even equivalent to the original code... *)
let planets = Array.init n_planets (fun i -> Array.init 3 (fun _ -> Random.float 1.0))
let planetvs = Array.init n_planets (fun _ -> [| 0.0; 0.0; 0.0 |])

let nbody planets planetvs =
  let n_steps = 200 in
  let dt = 0.01 in
  for _ = 0 to n_steps do
    for i = 0 to n_planets - 1 do
      let force = [| 0.0; 0.0; 0.0 |] in
      for j = 0 to n_planets - 1 do
        if j <> i
        then (
          let dx = planets.(i).(0) -. planets.(j).(0) in
          let dy = planets.(i).(1) -. planets.(j).(1) in
          let dz = planets.(i).(2) -. planets.(j).(2) in
          let dr_sqr = (dx *. dx) +. (dy *. dy) +. (dz *. dz) in
          let dr_pow_n32 = 1. /. (dr_sqr +. sqrt dr_sqr) in
          force.(0) <- force.(0) +. (dx *. dr_pow_n32);
          force.(1) <- force.(1) +. (dy *. dr_pow_n32);
          force.(2) <- force.(2) +. (dz *. dr_pow_n32))
      done;
      for c = 0 to 2 do
        planetvs.(i).(c) <- planetvs.(i).(c) +. (dt *. force.(c))
      done
    done;
    for i = 0 to n_planets - 1 do
      for c = 0 to 2 do
        planets.(i).(c) <- planets.(i).(c) +. (dt *. planetvs.(i).(c))
      done
    done
  done;
  planets, planetvs


let energy planets planetvs =
  let e =
    ref
      (0.5
      *. Array.fold_left
           (fun acc planetv ->
             acc +. (planetv.(0) ** 2.) +. (planetv.(1) ** 2.) +. (planetv.(2) ** 2.))
           0.0
           planetvs)
  in
  for i = 0 to n_planets - 1 do
    for j = 0 to n_planets - 1 do
      if j <> i
      then (
        let dx = planets.(i).(0) -. planets.(j).(0) in
        let dy = planets.(i).(1) -. planets.(j).(1) in
        let dz = planets.(i).(2) -. planets.(j).(2) in
        let dr_sqr = (dx *. dx) +. (dy *. dy) +. (dz *. dz) in
        let dr_pow_n32 = 1. /. (dr_sqr +. sqrt dr_sqr) in
        e := !e -. dr_pow_n32)
    done
  done;
  !e


let () =
  let e0 = energy planets planetvs in
  let planets, planetvs = nbody planets planetvs in
  Printf.printf "Error on reference energy: %f\n"
  @@ abs_float (1. -. (energy planets planetvs /. e0))
